### PR TITLE
fix for vs2015 build

### PR DESCRIPTION
**Pull request by sailfish009**
this fix enable build application without error with libuv.lib

### CODE REVIEW

File: include/uv-win.h [hunk 1]
```diff
@@ -647,3 +647,17 @@ RB_HEAD(uv_timer_tree_s, uv_timer_s);
 #ifndef X_OK
 #define X_OK 1
 #endif
+
+#if _MSC_VER >= 1900
+#undef stdin
```
Comment (saghul) [inline include/uv-win.h]: why is this needed? we can already compile libuv with MSVC 2015.

File: include/uv-win.h [hunk 2]
```diff
@@ -647,3 +647,17 @@ RB_HEAD(uv_timer_tree_s, uv_timer_s);
...
+#undef stdout
+#undef stderr
+_CRTIMP extern FILE _iob[];
+#define stdin   _iob
+#define stdout  (_iob+1)
+#define stderr  (_iob+2)
+#endif
+
+#pragma comment(lib, "IPHLPAPI.lib")
```
Comment (saghul) [inline include/uv-win.h]: this would break the MinGW build.

File: test/test-pipe-getsockname.c [hunk 3]
```diff
@@ -231,7 +231,7 @@ TEST_IMPL(pipe_getsockname_blocking) {
   len1 = sizeof buf1;
   r = uv_pipe_getsockname(&pipe_client, buf1, &len1);
   ASSERT(r == 0);
-  ASSERT(buf1[len1 - 1] != 0);
+  ASSERT(len1 == 0);  /* It's an annonymous pipe. */
```
Comment (saghul) [inline test/test-pipe-getsockname.c]: unrelated change

File: test/test-pipe-getsockname.c [hunk 4]
```diff
@@ -240,7 +240,7 @@ TEST_IMPL(pipe_getsockname_blocking) {
   len2 = sizeof buf2;
   r = uv_pipe_getsockname(&pipe_client, buf2, &len2);
   ASSERT(r == 0);
-  ASSERT(buf2[len2 - 1] != 0);
+  ASSERT(len2 == 0);  /* It's an annonymous pipe. */
```
Comment (saghul) [inline test/test-pipe-getsockname.c]: ditto

**Comment by saghul** [comment]
Can you elaborate on what doesn't work for you?

File: include/uv-win.h (hunk 1 above)
Comment (sailfish009) [inline include/uv-win.h]: with vs2015 update 3 or later version,

application compile failed.
unresolved external symbol __imp____iob_func referenced in ...

Comment (saghul) [inline include/uv-win.h]: how are you building libuv? please paste the full build output.

Comment (sailfish009) [inline include/uv-win.h]: just ran vcbuild.bat, libuv.lib built without any problem.

but if i compile application with libuv.lib, it failed to build

Comment (sailfish009) [inline include/uv-win.h]: for your information, here is some related link for the error:
http://stackoverflow.com/questions/30412951/unresolved-external-symbol-imp-fprintf-and-imp-iob-func-sdl2

Comment (bnoordhuis) [inline include/uv-win.h]: It sounds like you're not linking in the libraries that libuv needs.  From uv.gyp:
```
          'link_settings': {
            'libraries': [
              '-ladvapi32',
              '-liphlpapi',
              '-lpsapi',
              '-lshell32',
              '-luser32',
              '-luserenv',
              '-lws2_32'
            ],
          },
```

Comment (sailfish009) [inline include/uv-win.h]: python 2.7.12 installed, run vcbuild.bat in VS2015 x86 command prompt.

Comment (saghul) [inline include/uv-win.h]: I still don't understand what you mean. Maybe paste some output. But I don't think there is a problem with libuv. We already build 2 applications linked with libuv: run-tests.exe and run-benchmarks.exe, and they build just fine. You are probably missing linking with some libraries, as @bnoordhuis pointed out.

Comment (sailfish009) [inline include/uv-win.h]: i have create simple console application in vs2015,

and copy libuv.lib and header files to application source directory.

try to compile console application but show some compile error,
so i have to rebuild libuv.lib with my fix, that's all.

the source code is simple:

int main()
{
  uv_loop_t *loop = (uv_loop_t *)malloc(sizeof(uv_loop_t));
  uv_loop_init(loop);
  printf("Now quitting.\n");
  uv_run(loop, UV_RUN_DEFAULT);
  uv_loop_close(loop);
  free(loop);
  return 0;
}

Comment (saghul) [inline include/uv-win.h]: How are you linking your application? You have to link the libraries Ben pointed out.

Comment (sailfish009) [inline include/uv-win.h]: you are right, link these libs solved the problem.
but application side, this is little inconvenience.

#pragma comment(lib, "advapi32.lib")
#pragma comment(lib, "iphlpapi.lib")
#pragma comment(lib, "psapi.lib")
#pragma comment(lib, "shell32.lib")
#pragma comment(lib, "user32.lib")
#pragma comment(lib, "userenv.lib")
#pragma comment(lib,"ws2_32.lib")

Comment (saghul) [inline include/uv-win.h]: Feel free to submit a documentation PR, but i'm closing this one as it's not a bug.

Comment (sailfish009) [inline include/uv-win.h]: OOP, my mistake! 
This seems to be bug. 
I rechecked the case, here is compile error output:
i have created another simple console application,  compile error again,
link above libs does not help.

#include <stdio.h>
#include <stdlib.h>
#include "uv.h"

#pragma comment(lib, "libuv.lib")

#pragma comment(lib, "advapi32.lib")
#pragma comment(lib, "iphlpapi.lib")
#pragma comment(lib, "psapi.lib")
#pragma comment(lib, "shell32.lib")
#pragma comment(lib, "user32.lib")
#pragma comment(lib, "userenv.lib")
#pragma comment(lib,"ws2_32.lib")


int main()
{
...




1>------ Build started: Project: ConsoleApplication1, Configuration: Debug Win32 ------
1>  stdafx.cpp
1>  ConsoleApplication1.cpp
1>LINK : warning LNK4098: defaultlib 'LIBCMTD' conflicts with use of other libs; use /NODEFAULTLIB:library
1>libuv.lib(uv-common.obj) : error LNK2019: unresolved external symbol _fprintf referenced in function _uv__print_handles
1>libuv.lib(error.obj) : error LNK2001: unresolved external symbol _fprintf
1>libuv.lib(error.obj) : error LNK2019: unresolved external symbol ___iob_func referenced in function _uv_fatal_error
1>libuv.lib(util.obj) : error LNK2019: unresolved external symbol __snwprintf referenced in function _uv_cpu_info
1>libuv.lib(fs-event.obj) : error LNK2001: unresolved external symbol __snwprintf
1>libuv.lib(fs.obj) : error LNK2001: unresolved external symbol __fmode
1>libuv.lib(snprintf.obj) : error LNK2019: unresolved external symbol _vsnprintf_s referenced in function _snprintf
1>MSVCRTD.lib(vsnprintf_s.obj) : error LNK2001: unresolved external symbol _vsnprintf_s
1>libuv.lib(snprintf.obj) : error LNK2019: unresolved external symbol __vscprintf referenced in function _snprintf
1>libuv.lib(inet.obj) : error LNK2019: unresolved external symbol _sprintf referenced in function _inet_ntop6
1>MSVCRTD.lib(vsnprintf_s.obj) : error LNK2001: unresolved external symbol __vsnprintf_s
1>C:\ConsoleApplication1\Debug\ConsoleApplication1.exe : fatal error LNK1120: 8 unresolved externals
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

Comment (bnoordhuis) [inline include/uv-win.h]: LNK4098 means you're mixing debug/non-debug or single-threaded/multi-threaded CRT libraries.  Like @saghul already said, that's not an issue with libuv itself and this issue tracker really is not the right place to discuss it.  If you want to pursue this, please move it to the mailing list.

Comment (sailfish009) [inline include/uv-win.h]: there is some change in latest vs2015. here is link which is included my previous link:
https://msdn.microsoft.com/en-us/library/bb531344.aspx#BK_CRT

Comment (sailfish009) [inline include/uv-win.h]: http://stackoverflow.com/questions/32418766/c-unresolved-external-symbol-sprintf-and-sscanf-in-visual-studio-2015

this procedure works for me finally.

add legacy_stdio_definitions.lib to uv.sln in vs2015 AND application.
libuv property in vs2015 -> change platform toolset v100(visual studio 2010) to v140(visaul studio 2015)
clean and rebuild both libuv and application.

Comment (saghul) [inline include/uv-win.h]: Looks like you run vcbuild.bat which used Visudl Studio 2010 and then tried to use it from 2015. Nodejs does not run into any of these problems...

Comment (sailfish009) [inline include/uv-win.h]: i ran vcbuild.bat in vs2015 x86 native command prompt.
yes i have vs2010, too. but i didn't run vcbuild.bat in vs2010 prompt.
and vs2015 break change like this inline function, etc.
